src/psyclone/tests/psyir/tools/definition_use_chains_test.py

```diff
@@ -0,0 +1,755 @@
+!x = x + 1.0
+y = exp(x**2)
+end subroutine bar
+"""
+
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Routine)[0]
+    # Creating use chain for the a in c = a + 1.0
+    chains = DefinitionUseChain(
+        routine.children[0].children[1].children[0], [routine]
+    )
+    reaches = chains.find_forward_accesses()
+    # We find 3 results
+    # the a in e = a**2
+    # the a in c = d * a
+    # The call bar(c, b) as a isn't local and we can't guarantee its behaviour.
+    assert len(reaches) == 3
+    assert reaches[0] is routine.children[1].rhs.children[0]
+    assert reaches[1] is routine.children[4].rhs.children[1]
+    assert reaches[2] is routine.children[6]
+
+    # Create use chain for d in d = c + 2.0
+    chains = DefinitionUseChain(routine.children[3].lhs, [routine])
+    reaches = chains.find_forward_accesses()
+    # We should find 2 results
+    # c = D * a
+    # b = c + D
+    assert reaches[0] is routine.children[4].rhs.children[0]
+    assert reaches[1] is routine.children[5].rhs.children[1]
+    assert len(reaches) == 2
+
+    # Create use chain for c in c = d * a
+    chains = DefinitionUseChain(routine.children[4].lhs, [routine])
+    reaches = chains.find_forward_accesses()
+    # 2 results:
+    # b = C + d
+    # call bar(c, d)
+    assert len(reaches) == 2
+    assert reaches[0] is routine.children[5].rhs.children[0]
+    assert reaches[1] is routine.children[6].arguments[0]
+
+
+def test_definition_use_chain_find_forward_accesses_assignment(
+    fortran_reader,
+):
+
+    code = """
+    subroutine x()
+    integer :: a
+    a = 1
+    a = a * a
+    end subroutine
+    """
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Routine)[0]
+    # Start chain from A = 1
+    chains = DefinitionUseChain(routine.children[0].lhs)
+    reaches = chains.find_forward_accesses()
+    # We should find 3 results, all 3 references in
+    # A = A * A
+    assert len(reaches) == 3
+    assert reaches[0] is routine.children[1].rhs.children[0]
+    assert reaches[1] is routine.children[1].rhs.children[1]
+    assert reaches[2] is routine.children[1].lhs
+
+
+def test_definition_use_chain_find_forward_accesses_ifelse_example(
+    fortran_reader,
+):
+    """Functionality test for the find_forward_accesses routine. This
+    tests the behaviour when there is an if/else block."""
+
+    code = """
+    subroutine x()
+    integer :: a, b, c, d, e, f
+    a = 1
+    b = a + c
+    if ( d > e) then
+        a = 3
+    else
+        a = 4
+    end if
+    b = a + d
+    end subroutine"""
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.walk(Routine)[0]
+    # Start the chain from a = 1.
+    chains = DefinitionUseChain(routine.children[0].lhs)
+    reaches = chains.find_forward_accesses()
+    # TODO For now the if statement doesn't kill the accesses,
+    # even though it will always be written to.
+    # FIXME Add issue to the previous comment
```

Comment:
Yes please, create and reference in the TODO the issue for this.

Reply:
Done.